enhancements/microshift/microshift-updateability-ostree.md

```diff
@@ -0,0 +1,1051 @@
+- embedded in MicroShift binary list of blocked "from" versions
+
+A general flow will have following form:
+1. If persisted version is missing, assume 4.13.
+1. If version of `microshift` binary is older than version in metadata, **refuse to start MicroShift**.
+1. If persisted version is on a list of blocked version migrations, **refuse to start MicroShift**.
+1. If binary is the same version as persisted in metadata, **no need for a data migration**.
+1. Otherwise attempt to migrate the data.
+
+
+### Open Questions [optional]
+
+#### Should 4.13 -> 4.14 be supported?
+- Even if we handle "no metadata, no backup, existing data, no next-boot-action" and make
+  a backup, upon rollback 4.13 won't be able to restore it, so it would try to use 4.14's
+  data.
+  - Is manual intervention acceptable (admin manually copying backup to data)?
+  - We most likely don't want to implement special case for that (`pre-run` would save 
+    that persist it was 4.13 previously and red script would restore before shutting down)
+
+#### If green/red info is not persisted, and it's not a first boot (/var/lib/microshift exists), what should we do?
+- Compare versions (metadata vs binary)
+  - Same - start cluster
+  - Different - means that both binaries implement the enhancement,
+    so this could be considered a bug.
+  - Metadata is also missing
+    - Assume data was created by 4.13, do a backup and follow data migration rules
+
+#### When should version metadata file be created
+- By the green script
+  - Existence of file would indicate that data belongs to healthy deployment
+    - What upon reboot? Should file be cleared to be later recreated by green script?
+      - Blind assumption `.version exists == healthy data` might be too implicit and lead
+        to wrong conclusions
+- On start of `microshift run` or shortly after with addition from green and red scripts
+  - `{ "version": "X.Y.Z", "status": "unknown"}` would be created at first
+  - Green script would change `.status` to `healthy`
+  - Red script would change `.status` to `unhealthy`
+  - Pros:
+    - Explicit
+  - Cons:
+    - More complicated schema - weak con
+
+#### How should `microshift pre-run` be executed?
+- `microshift.service` - `ExecStartPre`
+  - No need to add new systemd service files.
+  - It will run on each `systemctl restart microshift` which is not desirable (will it run when systemd restarts MicroShift?)
+- `microshift-pre-run.service`
+  - Running on boot, just once, before `microshift.service`
+  - Not repeated on MicroShift restart
+  - New service file
+
+#### Should MicroShift healthcheck check and log version skew problems so it's easier to debug?
+- Why not?
+
+#### Should backups be kept only for deployments present in ostree command?
+- Is it possible that deployment can be reintroduced?
+  - I.e. it will have the same id and admin might want to rollback to it?
+
+#### Data exists, action is "restore", but backup does not exists
```

Comment:
This problem may be complex enough that waiting to approve this enhancement until we have a solution would prevent us from starting to build the other parts we need. Do you anticipate the solution to this needing fundamental changes in the approach? Or would it be more likely to be a refinement? Could we afford to set it aside to solve in the next sprint, after we have some experience building the early pieces of the backup and rollback system?

Reply:
I did a little thought experiment how "restore without backup" would look like with symlink based approach. PTAL https://gist.github.com/pmtk/12d7e99dca90fdeb461cfad2178b9c3b

It's quite interesting and we might at least consider it given early stages of implementation.
Also, it doesn't affect recoverability in "1st boot - OK, reboot, 2nd boot, backup OK, device unhealthy" scenario because
- 1st deployment wasn't technically staged, so greenboot didn't set up `boot_counter`
- 2nd deployment - if the first boot is okay, then greenboot sets `boot_success=1` and clears `boot_counter` so manual intervention is needed anyway

Reply:
I think we concluded in [this other thread](https://github.com/openshift/enhancements/pull/1387/files#r1181109156) that we do not want to use the "restore without backup" approach, right?